service/storage_service.cc

```diff
@@ -1855,6 +1855,33 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
     auto advertise = gms::advertise_myself(!replacing_a_node_with_same_ip);
     co_await _gossiper.start_gossiping(generation_number, app_states, advertise);
 
+    if (should_bootstrap()) {
+        // Wait for NORMAL state handlers to finish for existing nodes now, so that connection dropping
+        // (happening at the end of `handle_state_normal`: `notify_joined`) doesn't interrupt
+        // group 0 joining or repair. (See #12764, #12956, #12972, #13302)
+        //
+        // But before we can do that, we must make sure that gossip sees at least one other node
+        // and fetches the list of peers from it; otherwise `wait_for_normal_state_handled_on_boot`
+        // may trivially finish without waiting for anyone.
+        co_await _gossiper.wait_for_live_nodes_to_show_up(2);
+
+        auto ignore_nodes = ri
+                ? parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace(), get_token_metadata())
+                // TODO: specify ignore_nodes for bootstrap
+                : std::unordered_set<gms::inet_address>{};
+        auto sync_nodes = co_await get_nodes_to_sync_with(ignore_nodes);
+        if (ri) {
+            sync_nodes.erase(ri->address);
+        }
+
+        if (!_raft_topology_change_enabled) {
```

Comment:
Why not skip the entire outer 'if' in case of topology over raft? `get_nodes_to_sync_with` returns nodes that have normal status in gossiper anyway and this will be none eventually.

Reply:
Hm, what about waiting for nodes to be UP? We don't want to do it in raft-topology mode? It's still useful for ensuring that repair will not fail because gossiper thinks that a node is down.

> get_nodes_to_sync_with returns nodes that have normal status in gossiper anyway and this will be none eventually.

We could modify it to obtain nodes from topology state machine in raft-topology mode

Reply:
> > get_nodes_to_sync_with returns nodes that have normal status in gossiper anyway and this will be none eventually.
> 
> We could modify it to obtain nodes from topology state machine in raft-topology mode

But this can only by done much later. Here ewe do not have topology yet.

Reply:
Right... so we should do the "wait for UP" after we setup group 0...

But can't we leave it as it is for now? I feel like we're constantly trying to predict the future here and solve all (current and potential future) problems at once, and it's pretty annoying.

Reply:
> Right... so we should do the "wait for UP" after we setup group 0...
> 

Not even that since we are planning to move joining the cluster to the coordinator. It can be done just before streaming in the raft mode.

> But can't we leave it as it is for now? I feel like we're constantly trying to predict the future here and solve all (current and potential future) problems at once, and it's pretty annoying.

We do not try to predict the future. The code is definitely legacy specific and should be skipped  in topology over raft mode. I already added a bunch of ifs like that. If we add legacy code we need to run it conditionally from the start.

Reply:
> We do not try to predict the future. The code is definitely legacy specific and should be skipped in topology over raft mode. I already added a bunch of ifs like that. If we add legacy code we need to run it conditionally from the start.

The `if`s you added (`if (_raft_topology_enabled)`) were necessary for raft-topology mode to work.

In contrast, wait-for-UP code currently works in raft-topology mode too. It's not necessary to condition it as this point. Similarly we keep gossiping node statuses in raft-topology mode. It will be much simpler for the developer who disables gossiping node statuses in raft-topology mode to also move the wait-for-UP code to the appropriate place - they'll be able to test this, in contrast I have to predict that what I do will also work in the future after we remove gossiper node statuses.


> Not even that since we are planning to move joining the cluster to the coordinator.

Now we're trying to predict the future (how the code will be structured after we move joining to coordinator) :(

> It can be done just before streaming in the raft mode.

I think it's simplest for me to address your current concern and at the same time avoid too much clairvoyance, by moving the wait-for-UP code after `setup_group0` call.

Reply:
> > We do not try to predict the future. The code is definitely legacy specific and should be skipped in topology over raft mode. I already added a bunch of ifs like that. If we add legacy code we need to run it conditionally from the start.
> 
> The `if`s you added (`if (_raft_topology_enabled)`) were necessary for raft-topology mode to work.

Not necessary.  Look at this one for instance: fd1b1279c4324bc06a2d7710edc55546e72b93a5

> 
> In contrast, wait-for-UP code currently works in raft-topology mode too. It's not necessary to condition it as this point. Similarly we keep gossiping node statuses in raft-topology mode. It will be much simpler for the developer who disables gossiping node statuses in raft-topology mode to also move the wait-for-UP code to the appropriate place - they'll be able to test this, in contrast I have to predict that what I do will also work in the future after we remove gossiper node statuses.

I disagree. It will be a hassle since the code will stop working without nobody even notice that it stopped working. It is better to clearly distinguish raft code paths and gossiper code paths and not intermix them if not really needed.

> 
> > Not even that since we are planning to move joining the cluster to the coordinator.
> 
> Now we're trying to predict the future (how the code will be structured after we move joining to coordinator) :(

We already designed it and hopefully work on it. This is the present. Anyway doing it just before streaming would be correct for both cases.

> 
> > It can be done just before streaming in the raft mode.
> 
> I think it's simplest for me to address your current concern and at the same time avoid too much clairvoyance, by moving the wait-for-UP code after `setup_group0` call.

How will it help? I do not want this code to silently stop doing what it suppose to do when we remove the hack that adds nodes in normal state to gossiper in raft mode.

Reply:
> Anyway doing it just before streaming would be correct for both cases.

We need to do it before doing it the CDC write too.

I don't know where to put the call so it works for both cases.

In raft-topology mode streaming is done "somewhere here":
```cpp
        if (is_replacing()) {
            assert(raft_replace_info);
            co_await raft_replace(*raft_server, raft_replace_info->raft_id, raft_replace_info->ip_addr);
        } else {
            co_await raft_bootstrap(*raft_server);
        }

        // Wait until we enter one of the final states
        co_await _topology_state_machine.event.when([this, raft_server] {
            return _topology_state_machine._topology.normal_nodes.contains(raft_server->id()) ||
            _topology_state_machine._topology.left_nodes.contains(raft_server->id());
        });
```

and in gossiper mode it's done "later", in `bootstrap(` function which is not called in raft-topology mode - we return in the `if (raft_server)` branch.

So I think it would make most sense to put the wait before the `if (raft_server)` branch, i.e. somewhere around here:
```cpp
    co_await _group0->setup_group0(_sys_ks.local(), initial_contact_nodes, raft_replace_info, *this, qp, _migration_manager.local(), cdc_gen_service);

    raft::server* raft_server = co_await [this] () -> future<raft::server*> {
        if (!_raft_topology_change_enabled) {
            co_return nullptr;
        } else if (_sys_ks.local().bootstrap_complete()) {
            auto [upgrade_lock_holder, upgrade_state] = co_await _group0->client().get_group0_upgrade_state();
            co_return upgrade_state == group0_upgrade_state::use_post_raft_procedures ? &_group0->group0_server() : nullptr;
        } else {
            auto upgrade_state = (co_await _group0->client().get_group0_upgrade_state()).second;
            if (upgrade_state != group0_upgrade_state::use_post_raft_procedures) {
                on_internal_error(slogger, "raft topology: cluster not upgraded to use group 0 after setup_group0");
            }
            co_return &_group0->group0_server();
        }
    } ();

    auto schema_change_announce = _db.local().observable_schema_version().observe([this] (table_schema_version schema_version) mutable {
        _migration_manager.local().passive_announce(std::move(schema_version));
    });
    _listeners.emplace_back(make_lw_shared(std::move(schema_change_announce)));
    co_await _gossiper.wait_for_gossip_to_settle();
    co_await _feature_service.enable_features_on_join(_gossiper, _sys_ks.local());

    set_mode(mode::JOINING);

    if (raft_server) { // Raft is enabled. Check if we need to bootstrap ourself using raft
```

so putting it after `setup_group0()` looks sensible.

> How will it help? I do not want this code to silently stop doing what it suppose to do when we remove the hack that adds nodes in normal state to gossiper in raft mode.

Then your problem is not the location of the call, it's about modifying `get_nodes_to_sync_with` to take the nodes from raft topology in raft mode, no?

Reply:
> > Anyway doing it just before streaming would be correct for both cases.
> 
> We need to do it before doing it the CDC write too.

We can have it as a separate saga command that coordinator sends to a bootstrapping node.

> 
> I don't know where to put the call so it works for both cases.
> 

That's my point. There is no such place.

> In raft-topology mode streaming is done "somewhere here":
> 

It is done far from here in cmd handling and is alredy complete at the end fof this snippet.

> ```c++
>         if (is_replacing()) {
>             assert(raft_replace_info);
>             co_await raft_replace(*raft_server, raft_replace_info->raft_id, raft_replace_info->ip_addr);
>         } else {
>             co_await raft_bootstrap(*raft_server);
>         }
> 
>         // Wait until we enter one of the final states
>         co_await _topology_state_machine.event.when([this, raft_server] {
>             return _topology_state_machine._topology.normal_nodes.contains(raft_server->id()) ||
>             _topology_state_machine._topology.left_nodes.contains(raft_server->id());
>         });
> ```
> 
> and in gossiper mode it's done "later", in `bootstrap(` function which is not called in raft-topology mode - we return in the `if (raft_server)` branch.
> 
> So I think it would make most sense to put the wait before the `if (raft_server)` branch, i.e. somewhere around here:
> 
> ```c++
>     co_await _group0->setup_group0(_sys_ks.local(), initial_contact_nodes, raft_replace_info, *this, qp, _migration_manager.local(), cdc_gen_service);
> 
>     raft::server* raft_server = co_await [this] () -> future<raft::server*> {
>         if (!_raft_topology_change_enabled) {
>             co_return nullptr;
>         } else if (_sys_ks.local().bootstrap_complete()) {
>             auto [upgrade_lock_holder, upgrade_state] = co_await _group0->client().get_group0_upgrade_state();
>             co_return upgrade_state == group0_upgrade_state::use_post_raft_procedures ? &_group0->group0_server() : nullptr;
>         } else {
>             auto upgrade_state = (co_await _group0->client().get_group0_upgrade_state()).second;
>             if (upgrade_state != group0_upgrade_state::use_post_raft_procedures) {
>                 on_internal_error(slogger, "raft topology: cluster not upgraded to use group 0 after setup_group0");
>             }
>             co_return &_group0->group0_server();
>         }
>     } ();
> 
>     auto schema_change_announce = _db.local().observable_schema_version().observe([this] (table_schema_version schema_version) mutable {
>         _migration_manager.local().passive_announce(std::move(schema_version));
>     });
>     _listeners.emplace_back(make_lw_shared(std::move(schema_change_announce)));
>     co_await _gossiper.wait_for_gossip_to_settle();
>     co_await _feature_service.enable_features_on_join(_gossiper, _sys_ks.local());
> 
>     set_mode(mode::JOINING);
> 
>     if (raft_server) { // Raft is enabled. Check if we need to bootstrap ourself using raft
> ```
> 
> so putting it after `setup_group0()` looks sensible.

How? In raft mode at this point we will not know an other nodes at all.

> 
> > How will it help? I do not want this code to silently stop doing what it suppose to do when we remove the hack that adds nodes in normal state to gossiper in raft mode.
> 
> Then your problem is not the location of the call, it's about modifying `get_nodes_to_sync_with` to take the nodes from raft topology in raft mode, no?

No, because in raft node at the place you are calling it we will not know topology yet, so we will not know other nodes.

Reply:
> How? In raft mode at this point we will not know an other nodes at all.
> No, because in raft node at the place you are calling it we will not know topology yet, so we will not know other nodes.

But we did a read barrier here so what do you mean? We have topology.

Unless you mean it will be different after we completely change this code to do node joining in a completely different way. And that's my problem - it's hard for me at this point to solve your problem, because I don't know how the code will look in the future... You're trying to solve every problem at once, instead of doing incremental development. Which is preventing useful CI stability fixes from being merged right now.

Reply:
> That's my point. There is no such place.

**Currently** there is such a place and it's exactly where I'm saying.
What you're saying is that **in the future** this place won't work. But what should I do? It's impossible for me to create a fix that way.

Reply:
> > In one week future?
> 
> Heh, sure

We do work on it, no?

Reply:
> 
> I guess I can leave it half-fixed (gossiper-mode only) and say "Refs:"...

You can document what is missing there as well.